pkg/api/defaults-custom-cloud-profile.go

```diff
@@ -3,34 +3,110 @@
 
 package api
 
-import "github.com/Azure/aks-engine/pkg/helpers"
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+	"time"
 
-func (p *Properties) setCustomCloudProfileDefaults() {
+	"github.com/Azure/aks-engine/pkg/helpers"
+	"github.com/Azure/go-autorest/autorest/azure"
+)
 
+func (cs *ContainerService) setCustomCloudProfileDefaults() {
+	p := cs.Properties
 	if p.IsAzureStackCloud() {
+		p.CustomCloudProfile.AuthenticationMethod = helpers.EnsureString(p.CustomCloudProfile.AuthenticationMethod, ClientSecretAuthMethod)
+		p.CustomCloudProfile.IdentitySystem = helpers.EnsureString(p.CustomCloudProfile.IdentitySystem, AzureADIdentitySystem)
+		p.CustomCloudProfile.DependenciesLocation = DependenciesLocation(helpers.EnsureString(string(p.CustomCloudProfile.DependenciesLocation), AzureStackDependenciesLocationPublic))
+		p.SetCustomAzureCloudSpec()
+		err := cs.SetCustomCloudProfileEnvironment()
+		if err != nil {
+			log.Fatalf("Failed to set environment - %s", err)
```

Comment:
Instead of logging errors and exiting here, make the function return type an error and return the error. The only place that should exit and log the error is the original caller, main.go.

Same comment for all the other log.Fatalf calls in the code.

Reply:
I agree: this should be a `log.Errorf` instead, and the error should be returned. If we have code that calls `os.Exit`—which `Fatalf` does—it basically can't be unit-tested. (See also #536 for an example of how we've tried to get rid of `Fatalf`.)